### DISCUSSION TITLE
Doubts about display-sync-status --dry-run

### DISCUSSION THREAD

**Discussion by jrpelegrina** [abraunegg/onedrive#2008] [General]
Hello
First of all congratulations for the great job you are doing with the client for OneDrive. It's fantastic.
I wanted to ask you about how the display-sync-status command works when run with --dry-run.

The situation is the following:
I am using version 2.4.19 of the client offered by OpenSuse. I have the client running through a systemd service in monitor mode on an Ubuntu 20.04 machine.

If I make a change in the OneDrive cloud, the change is correctly reflected in the local folder I use to sync, the log tells me that everything is synced, but if I run a display-sync-status ---dry-run it tells me that there is content pending to be synchronized (which corresponds to the change made in the cloud). If after a few minutes I repeat the command, the response remains the same.

Only if I stop the service and therefore onedrive is stopped, the command display-sync-status --dry-run tells me that all the content is synchronized.

Is this normal behavior?

All the best

**Comment by abraunegg** [comment]
@jrpelegrina 
> Is this normal behavior?

This is probably more of a usage quirk than anything else .. let me try and explain why you are seeing this behavior.

**Some background history:**

In v2.4.17 and below, we had several *bugs* related to especially Ubuntu platforms - and where the package for the platform would create a systemd service automatically - so that, the moment you authenticated the application, the service would start and start downloading content. This was not a 'onedrive' bug - but a packaging issue.

What folk running Ubuntu would do however - would then also try running the application manually, and/or setup their own systemd service using the same configuration - thus potentially having 2 or 3 active application instances running against the same configuration. From time-to-time these running instances would collide, and generate a application error / database access error / database I/O error / database corruption and the application would fault out.

In v2.4.18 the packaging issue was *fixed*, but, if you had the older systemd config and/or still tried to run multiple instances of the application - this would still cause an issue, so in v2.4.18 all the backend database access was reconfigured to 'exlusive' mode - meaning - single process per configuration - which is the right approach to prevent any database issue that was seen with v2.4.17 and earlier.

In 2.4.19 a major disk write bug was fixed (this bug had been in the client for 5 years), which was essentially quite probably a leading contributor also to the database access issues / corruption issues seen with v2.4.17 and earlier if multiple instances were trying to re-write the database at the same time. This change also changed when the 'in-flight' changes from OneDrive are flushed out to full database - to reduce the constant disk writes every check of OneDrive for new data (as defined by 'monitor_interval') and only flushes this data out to the database once the integrity check is complete (as defined by 'monitor_fullscan_frequency').

**Attempt to answer your question:**

So when you are using `--display-sync-status ---dry-run` you are making a *copy* of the `items.sqlite3` file .. so *if* the changes have not been flushed, then, you are making a copy of the prior state, before you enabled the systemd service, and in-between the committal of data to disk once the integrity check is done. This is why it is reporting that there are changes to apply. All of those 'in-between' changes will not be in the current `items.sqlite3` file.

When you stop the systemd service, you flush this temp data to `items.sqlite3` - so this is why when you re-run the command everything is in-sync.

Now - if you just try and use `--display-sync-status` and your systemd service is running - this will most likely generate an error with v2.4.19 - citing that the application is already running .. which it is .. via systemd


Let me have a think about this a little further and see if there is solution to this quirk that can be adopted.

**Reply by rrodrigueznt** [reply]
Please, when running a  `onedrive v2.4.20` on a fully updated CentOS Stream 8 server should I expect that `--display-sync-status --dry-run` gives me the righth information about the synchronisation status?

Thanks!

**Reply by abraunegg** [reply]
@rrodrigueznt 
When using `--dry-run` - because of how the client operates (by making a *copy* of the operational sync state database), the `--display-sync-status` will *mostly* reflect an out-of-sync state.

This is because of the in memory operations that the client utilises from a performance perspective are written to disk (the actual file that `--dry-run` copies) at the completion of 'monitor_fullscan_frequency' have most likely not yet been committed to disk as per the detailed explaination above.

The only way to obtain the accurate `--display-sync-status` response, is to execute the command against a non-running instance.

**Reply by rrodrigueznt** [reply]
Why not flush in-memory data out to the database when `--dry-run` is used? Wouldn't this give us the best information about the synchronization state?

**Reply by abraunegg** [reply]
> Why not flush in-memory data out to the database when `--dry-run` is used? Wouldn't this give us the best information about the synchronization state?

It simply cannot be done - because it is a separate running process ...

* Process 1 = active, locked DB state, resident memory contains changes or committed items, that have not been committed to disk
* Process 2 = active, using process 1 DB copy to get 'last known committed state' as starting point

Think of it like a 2 lane highway, with 2 cars travelling the same speed, same direction. You want to change the radio station in the second car but you have no way reach in and change the radio station.

**Reply by rrodrigueznt** [reply]
Please, should you stop/deactivate process 1 to make the DB copy used in process 2? I'm afraid I'm failing to understand something obvious related with SQLite functioning. Thanks!

**Reply by abraunegg** [reply]
Its called DB locking - this is why, when using `--dry-run` - it is using a DB copy - so changes - destructive / data changing - are done on that copy in a non-destructive way.

If you stop process 1, it will commit the data to disk, making the local db cache correct and true as all inflight records are written out. They are also written out post the integrity check.

**Reply by rrodrigueznt** [reply]
Thanks for your patience. I will keep struggling to understand the whole thing as much as I can with my level of skills! Please, just one more question: does the --dry-run copy of the DB persist? In the lines below, you will see a `items-dryrun.sqlite3` file stored in my system that disappears after a new `--display-sync-status --dry-run` was done with the same configuration.

```
(base) [rjr@e220i059h057w onedrive_igfae-dir]$ onedrive --confdir="~/.config/onedrive_igfae-dir" --display-sync-status
Configuration file successfully loaded
config file has been updated, checking if --resync needed
Configuring Global Azure AD Endpoints

ERROR: onedrive application is already running - check system process list for active application instances

(base) [rjr@e220i059h057w onedrive_igfae-dir]$ ls -l
total 5092
-rw-rw-r--. 1 rjr rjr      1780 Aug 15 10:52 config
-rw-r--r--. 1 rjr rjr    524288 Aug  6 10:26 items-dryrun.sqlite3
-rw-r--r--. 1 rjr rjr    524288 Aug 15 07:26 items.sqlite3
-rw-r--r--. 1 rjr users   32768 Aug 15 10:52 items.sqlite3-shm
-rw-r--r--. 1 rjr users 4124152 Aug 15 10:52 items.sqlite3-wal
-rw-------. 1 rjr rjr       929 Aug 15 10:32 refresh_token
(base) [rjr@e220i059h057w onedrive_igfae-dir]$ onedrive --confdir="~/.config/onedrive_igfae-dir" --display-sync-status --dry-run
Configuration file successfully loaded
DRY-RUN Configured. Output below shows what 'would' have occurred.
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
No pending remote changes - in sync
(base) [rjr@e220i059h057w onedrive_igfae-dir]$ ls -l
total 4580
-rw-rw-r--. 1 rjr rjr      1780 Aug 15 10:52 config
-rw-r--r--. 1 rjr rjr    524288 Aug 15 07:26 items.sqlite3
-rw-r--r--. 1 rjr users   32768 Aug 15 10:52 items.sqlite3-shm
-rw-r--r--. 1 rjr users 4124152 Aug 15 10:52 items.sqlite3-wal
-rw-------. 1 rjr rjr       929 Aug 15 10:32 refresh_token
(base) [rjr@e220i059h057w onedrive_igfae-dir]$ onedrive --confdir="~/.config/onedrive_igfae-dir" --display-sync-status
Configuration file successfully loaded
Configuring Global Azure AD Endpoints

ERROR: onedrive application is already running - check system process list for active application instances

(base) [rjr@e220i059h057w onedrive_igfae-dir]$ ls -l
total 4580
-rw-rw-r--. 1 rjr rjr      1780 Aug 15 10:52 config
-rw-r--r--. 1 rjr rjr    524288 Aug 15 07:26 items.sqlite3
-rw-r--r--. 1 rjr users   32768 Aug 15 10:52 items.sqlite3-shm
-rw-r--r--. 1 rjr users 4124152 Aug 15 10:52 items.sqlite3-wal
-rw-------. 1 rjr rjr       929 Aug 15 10:32 refresh_token
(base) [rjr@e220i059h057w onedrive_igfae-dir]$
```

Thanks!

**Reply by abraunegg** [reply]
@rrodrigueznt 
The dry-run copy should be cleaned up post usage on exit. The application also checks if it exists at startup, and, if it does, cleans it up as it should not exist if not being used.

What you are seeing is normal.

**Comment by jrpelegrina** [comment]
Thank you very much for answering my query so quickly and for the information provided.

I will be attentive to the possible news on this subject
